Modify llvm-gsymutil lookups to handle overlapping ranges correctly.

clayborg opened the pull request:
llvm-gsymutil allows address ranges to overlap. There was a bug where if we had debug info for a function with a range like [0x100-0x200) and a symbol at the same start address yet with a larger range like [0x100-0x300), we would randomly get either only information from the first or second entry. This could cause lookups to fail due to the way the binary search worked.

This patch makes sure that when lookups happen we find the first address table entry that can match an address, and also ensures that we always select the first FunctionInfo that could match. FunctionInfo entries are sorted such that the most debug info rich entries come first. And if we have two ranges that have the same start address, the smaller range comes first and the larger one comes next. This patch also adds the ability to iterate over all function infos with the same start address to always find a range that contains the address.

Added a unit test to test this functionality that failed prior to this fix and now succeeds.

clayborg committed:
Modify llvm-gsymutil lookups to handle overlapping ranges correctly.

llvm-gsymutil allows address ranges to overlap. There was a bug where if we had debug info for a function with a range like [0x100-0x200) and a symbol at the same start address yet with a larger range like [0x100-0x300), we would randomly get either only information from the first or second entry. This could cause lookups to fail due to the way the binary search worked.

This patch makes sure that when lookups happen we find the first address table entry that can match an address, and also ensures that we always select the first FunctionInfo that could match. FunctionInfo entries are sorted such that the most debug info rich entries come first. And if we have two ranges that have the same start address, the smaller range comes first and the larger one comes next. This patch also adds the ability to iterate over all function infos with the same start address to always find a range that contains the address.

Added a unit test to test this functionality that failed prior to this fix and now succeeds.

clayborg commented:
> > we had debug info for a function with a range like [0x100-0x200) and a symbol at the same start address yet with a larger range like [0x100-0x300)
> 
> Do you have a test case for where this happens? If clang/llvm are producing debug info like that I'd be concerned/want to look into fixing it.

This is Haskell debugging related where the compiler is producing debug info with a small sized `DW_TAG_subprogram` and the symbol table symbol has a much larger range. I don't believe clang supports Haskell. The `DW_AT_producer` in this case is:
```
DW_AT_producer    ("The Glorious Glasgow Haskell Compilation System 8.8.3")
```

dwblaikie commented:
> > > we had debug info for a function with a range like [0x100-0x200) and a symbol at the same start address yet with a larger range like [0x100-0x300)
> > 
> > 
> > Do you have a test case for where this happens? If clang/llvm are producing debug info like that I'd be concerned/want to look into fixing it.
> 
> This is Haskell debugging related where the compiler is producing debug info with a small sized `DW_TAG_subprogram` and the symbol table symbol has a much larger range. I don't believe clang supports Haskell. The `DW_AT_producer` in this case is:
> 
> ```
> DW_AT_producer    ("The Glorious Glasgow Haskell Compilation System 8.8.3")
> ```

ah, thanks for the details. Yeah, I can see how merging both the ELF symbol table and the DWARF description could end up with differences, possibly due to whether the alignment padding gets included in one description or the other - probably better that it not be included in either, but yeah, such is life.

ayermolo commented on llvm/lib/DebugInfo/GSYM/GsymReader.cpp:
```diff
@@ -253,49 +253,66 @@ GsymReader::getAddressIndex(const uint64_t Addr) const {
+      break; // Done with consecutive function info entries with same address.
+
+    // Address info offsets size should have been checked in parse().
+    auto AddrInfoOffset = AddrInfoOffsets[AddrIdx];
```
I realize it was auto in original code, but can we change it to actual type?

ayermolo commented on llvm/include/llvm/DebugInfo/GSYM/GsymReader.h:
```diff
@@ -266,6 +266,18 @@ class GsymReader {
     if (Iter == End || AddrOffset < *Iter)
       --Iter;
+
+    // GSYM files store the richest information first in the file, so always
```
What do you mean by richest?

clayborg committed:
Respond to feedback and fix dump issue.

Fixed a case where when we dumped an entire GSYM file we would dump the wrong FunctionInfo for a given address entry. This was because we would lookup a function info by address when dumping a specific address table entry. If the address table had multiple addresses with different ranges, it would always emit the one that was found via lookup by address.

clayborg committed:
Fix comment typos.

clayborg committed:
clang-format
